pkg/lib/cockpit-upload-helper.js

```diff
@@ -0,0 +1,164 @@
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2024 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+import cockpit from "cockpit";
+
+const BLOCK_SIZE = 16 * 1024;
+const TOTAL_FRAME_WINDOW = 128;
+/*
+ * Cockpit Upload helper - uploads a blob to the provided destination with an
+ * optional progress callback using the fsreplace1 channel.
+ *
+ * Example usage:
+ *
+ * const helper = UploadHelper("/tmp/file.txt")
+ * try {
+ *   const status = await helper.upload(file);
+ * } catch (exc) {
+ * }
+ *
+ * Uploads can be cancelled while in progress by calling `helper.cancel`,
+ * fsreplace1 will automatically clean up the temporary file.
+ */
+export class UploadHelper {
+    constructor(destination, onProgress, options) {
+        this.destination = destination;
+        this.progressCallback = onProgress;
+        this.outstanding = 0;
+
+        // Promise resolvers
+        this.resolveHandler = null;
+
+        options = options || {};
+        this.channel = cockpit.channel({
+            binary: true,
+            payload: "fsreplace1",
+            path: this.destination,
+            "send-acks": "frames",
+            ...options,
+        });
+        this.channel.addEventListener("control", this.on_control.bind(this));
+    }
+
+    // Private methods
+    on_control(event, message) {
+        if (message?.command === "ack") {
+            this.outstanding -= message.frames;
+
+            console.assert(this.outstanding >= 0, "outstanding went negative", this.outstanding);
+
+            if (this.resolveHandler) {
+                this.resolveHandler();
+                this.resolveHandler = null;
+            }
+        }
+    }
+
+    read_chunk(blob) {
+        return new Promise((resolve, reject) => {
+            const reader = new FileReader();
+            reader.onload = (event) => resolve(event.target.result);
+            reader.onerror = reject;
+            reader.readAsArrayBuffer(blob);
+        });
+    }
```

Comment:
I think this could be replaced with something like

```new Uint8Array(await blob.arrayBuffer());```

but I didn't test it...

Reply:
So how would that code even work? We resolve `event.target.result`. How would one `await` an arrayBuffer?